capellambse/decl.py

```diff
@@ -0,0 +1,404 @@
+            parent = model.by_uuid(parent.uuid)
+        for op_type, apply_op in _OPERATIONS.items():
+            try:
+                op = instruction.pop(op_type)
+            except KeyError:
+                continue
+            for promise, outcome in apply_op(promises, parent, op):
+                if isinstance(outcome, dict):
+                    deferred[promise].append(outcome)
+                else:
+                    if promise in promises:
+                        raise ValueError(
+                            f"promise_id defined twice: {promise.identifier}"
+                        )
+                    promises[promise] = outcome
+                    instructions.extend(deferred.pop(promise, ()))
+        if instruction:
+            keys = ", ".join(instruction)
+            raise ValueError(f"Unrecognized keys in instruction: {keys}")
+    if deferred:
+        raise UnfulfilledPromisesError(frozenset(deferred))
+
+
+def _operate_create(
+    promises: dict[Promise, capellambse.ModelObject],
+    parent: capellambse.ModelObject,
+    creations: dict[str, t.Any],
+) -> cabc.Generator[_OperatorResult, t.Any, None]:
+    for attr, value in creations.items():
+        if not isinstance(value, cabc.Iterable):
+            raise TypeError("values below `create:*:` must be lists")
+
+        yield from _create_complex_objects(promises, parent, attr, value)
+
+
+def _operate_delete(
+    promises: dict[Promise, capellambse.ModelObject],
+    parent: capellambse.ModelObject,
+    deletions: dict[str, t.Any],
+) -> cabc.Iterable[_OperatorResult]:
+    del promises
+
+    for attr, objs in deletions.items():
+        try:
+            target = getattr(parent, attr)
+        except AttributeError:
+            raise TypeError(
+                "Cannot delete object:"
+                f" {type(parent).__name__} has no attribute {attr!r}"
+            ) from None
+        if not isinstance(target, common.ElementList) or not isinstance(
+            objs, list
+        ):
+            delattr(parent, attr)
+            continue
+        if not isinstance(target, common.ElementListCouplingMixin):
+            raise TypeError(
+                "Cannot delete object:"
+                f" {type(parent).__name__}.{attr} is not model-coupled"
+            )
+        uuids = list(t.cast(cabc.Iterable[str], target.by_uuid))
+        for obj in objs:
+            if isinstance(obj, UUIDReference):
+                obj = obj.uuid
+            if not isinstance(obj, str):
+                raise TypeError("Values in `delete:*:` must be UUIDs")
+            try:
+                idx = uuids.index(obj)
+            except IndexError:
+                puuid = getattr(parent, "uuid", None)
+                raise ValueError(
+                    f"No object with UUID {obj!r} in {attr!r} of {puuid!r}"
+                ) from None
+            del uuids[idx]
+            del target[idx]
+
+    return ()
+
+
+def _operate_modify(
```

Comment:
IMO I'd like to have this public with a section about possible raised exceptions.